genicam: add 'Representation' property node support

sh0 opened the pull request:
This commit adds support for Representation property nodes which specify how integer and float values should be shown to user (e.g. as hex values).

From limited testing it seems that everything works, but testing/review would be much appreciated. Unit tests succeed, but I have not added extra tests for checking representation property.

I have used both `error` and `default_value` style functions. The `ArvGc*Interface` methods all have `error` parameters so I followed that with `get_representation` functions. However in `arv_gc_property_node_get_representation` I used `default_value` to match other property node functions. The result is that `get_representation` really never throws an error and silently returns `ARV_GC_REPRESENTATION_UNDEFINED` on errors. The default value for all nodes is actually `ARV_GC_REPRESENTATION_PURE_NUMBER`. Should we return this instead?

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcmaskedintregnode.c  1
- src/arvgcinteger.c  1
- src/arvgcfloatregnode.c  1
- src/arvgcintregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcmaskedintregnode.c  1
- src/arvgcinteger.c  1
- src/arvgcintregnode.c  1
- src/arvgcfloatregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcmaskedintregnode.c  1
- src/arvgcfloatregnode.c  1
- src/arvgcinteger.c  1
- src/arvgcintregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

sh0 commented:
Sorry for the post-pull changes - there were some minor bugs. I also have 'Representation' and 'Unit' property support for SwissKnife nodes ready to go. Also there is support for 'DisplayNotation' and 'DisplayPrecision' in yet another branch ready, but I'll wait this commit out first.

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcmaskedintregnode.c  1
- src/arvgcinteger.c  1
- src/arvgcfloatregnode.c  1
- src/arvgcintregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcinteger.c  1
- src/arvgcfloatregnode.c  1
- src/arvgcmaskedintregnode.c  1
- src/arvgcintregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

EmmanuelP commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgcmaskedintregnode.c  1
- src/arvgcinteger.c  1
- src/arvgcintregnode.c  1
- src/arvgcfloatregnode.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=5809860&bid=19018835)

sh0 commented:
Thank you for the review! I removed the error throwing code for arvgcfloat and arvgcinteger, squashed the commits and rebased on master.

It seems that in addition to `<Unit>` node we probably shouldn't report errors from lack of `<Inc>`, `<Min>` and `<Max>` nodes either. Maybe I'll try to patch this at some point.

The default value issue is still bugging me though. Currently `<Inc>`, `<Min>` and `<Max>` functions return default values as suggested by the standard (e.g. 1, `G_MININT64`, `G_MAXINT64`). The `<DisplayPrecision>` patch I have in progress will also return 6 as suggested by standard. At the same time all enum nodes return the `UNDEFINED` values instead of defaults in standards. Should we continue with this style or what do you think?

EmmanuelP commented:
Thanks.

Depending on the time you have, I would not mind some checks in the test suite for the new API. But it is really up to you, not a requirement.

Cheers.

sh0 commented:
I haven't gotten around to see how the test framework operates, but I'll keep it in mind to add some tests 👍
